### PR TITLE
Dataset size checks in preprocess_for_training

### PR DESCRIPTION
**Pull request by dantreiman**
Check dataset sizes after preprocessing, warn if validation or test set is too small to compute metrics.

### CODE REVIEW

File: ludwig/data/preprocessing.py [hunk 1]
```diff
@@ -1647,32 +1648,45 @@ def preprocess_for_training(
         with backend.storage.cache.use_credentials() if cached else contextlib.nullcontext():
             logger.debug("create training dataset")
             training_dataset = backend.dataset_manager.create(training_set, config, training_set_metadata)
-            if not len(training_set):
+            training_set_size = len(training_dataset)
+            if training_set_size == 0:
                 raise ValueError("Training data is empty following preprocessing.")
+            elif training_set_size < MIN_DATASET_SPLIT_ROWS:
+                raise ValueError(f"Training dataset has only {training_set_size} rows following preprocessing.")
```
Comment (arnavgarg1) [inline ludwig/data/preprocessing.py:1655]: Should we add that it needs a minimum of X rows as a part of the error message?

File: ludwig/data/preprocessing.py [hunk 2]
```diff
@@ -1647,32 +1648,45 @@ def preprocess_for_training(
...
                     logger.warning(
-                        "Encountered empty validation dataset. If this is unintentional, please check the "
-                        "preprocessing configuration."
+                        "Validation set empty. If this is unintentional, please check the preprocessing configuration."
+                    )
+                    validation_dataset = None
+                elif validation_set_size < MIN_DATASET_SPLIT_ROWS:
+                    logger.warning(
+                        f"Validation set ({validation_set_size}) too small to compute metrics. Evaluation of validation"
+                        " set will be skipped"
```
Comment (arnavgarg1) [inline ludwig/data/preprocessing.py:1670]: Would it help to tell the user the number of rows it has and the number of rows it should have to not skip validation?

File: ludwig/data/preprocessing.py [hunk 3]
```diff
@@ -1647,32 +1648,45 @@ def preprocess_for_training(
...
-                        "Encountered empty test dataset. If this is unintentional, please check the "
-                        "preprocessing configuration."
+                        "Test set empty. If this is unintentional, please check the preprocessing configuration."
                     )
                     test_dataset = None
+                elif test_set_size < MIN_DATASET_SPLIT_ROWS:
+                    logger.warning(
+                        f"Test set ({test_set_size}) too small to compute metrics. Evaluation of test set will be"
+                        " skipped"
+                    )
```
Comment (arnavgarg1) [inline ludwig/data/preprocessing.py:1688]: Same as above

File: ludwig/data/preprocessing.py (hunk 1 above)
Comment (dantreiman) [reply ludwig/data/preprocessing.py]: Good call, thanks!

File: ludwig/data/preprocessing.py [hunk 4]
```diff
@@ -1647,32 +1648,48 @@ def preprocess_for_training(
...
-                        "preprocessing configuration."
+                        "Test set empty. If this is unintentional, please check the preprocessing configuration."
                     )
                     test_dataset = None
+                elif test_set_size < MIN_DATASET_SPLIT_ROWS:
+                    logger.warning(
+                        f"Test set too small to compute metrics. Need at least {MIN_DATASET_SPLIT_ROWS} rows, got"
+                        f" {test_set_size} after preprocessing. Evaluation of test set will be skipped."
+                    )
+                    validation_dataset = None
```
Comment (arnavgarg1) [inline ludwig/data/preprocessing.py:1692]: Typo: This should probably be `test_dataset = None`?

Comment (dantreiman) [reply ludwig/data/preprocessing.py]: Copy/paste error, thanks for reading carefully!

**Review by arnavgarg1 — APPROVED** [review]
Looks good to me - would just wait for the tests to run since I have a feeling this may result in a few test failures where the number of examples being used is 25 instead of maybe like >= 50.

**Comment by dantreiman** [comment]
> Looks good to me - would just wait for the tests to run since I have a feeling this may result in a few test failures where the number of examples being used is 25 instead of maybe like >= 50.

You are right; test_preprocessing.py fails for exactly that reason.  Maybe it should only warn for small validation/test set but not set them to None as if they were empty?

**Pull request merged**